### PR TITLE
fix 32 char SSIDs

### LINKED ISSUE #6452: WiFi not working since 2024.11.0
### The problem

After updating, I cannot connect the ESPs to my wifi:

[08:12:32][D][wifi:497]: Found networks:
[08:12:32][I][wifi:541]: - '0619562e-bcbf-4bfc-97a8-b19e3c8f' [redacted]▂▄▆█
[08:12:32][D][wifi:542]:     Channel: 11
[08:12:32][D][wifi:543]:     RSSI: -76 dB
[08:12:32][I][wifi:541]: - '0619562e-bcbf-4bfc-97a8-b19e3c8f' [redacted]▂▄▆█
[08:12:32][D][wifi:542]:     Channel: 11
[08:12:32][D][wifi:543]:     RSSI: -69 dB
[08:12:32][I][wifi:541]: - '0619562e-bcbf-4bfc-97a8-b19e3c8f' [redacted]▂▄▆█
[08:12:32][D][wifi:542]:     Channel: 11
[08:12:32][D][wifi:543]:     RSSI: -58 dB
[08:12:38][I][wifi:313]: WiFi Connecting to '0619562e-bcbf-4bfc-97a8-b19e3c8f'...
[08:12:38][W][wifi:663]: WiFi network can not be found anymore.
[08:12:43][I][wifi:313]: WiFi Connecting to '0619562e-bcbf-4bfc-97a8-b19e3c8f'...
[08:12:43][W][wifi:663]: WiFi network can not be found anymore.
[08:12:43][I][wifi:313]: WiFi Connecting to '0619562e-bcbf-4bfc-97a8-b19e3c8f'...
[08:12:43][W][wifi:663]: WiFi network can not be found anymore.
[08:13:58][W][wifi_esp32:534][arduino_events]: Event: Disconnected ssid='0619562e-bcbf-4bfc-97a8-b19e3c8' reason='Probe Request Unsuccessful'

### Which version of ESPHome has the issue?

2024.11.0

### What type of installation are you using?

Home Assistant Add-on

### Which version of Home Assistant has the issue?

2024.11.2

### What platform are you using?

ESP32

### Board

esp32dev

### Component causing the issue

wifi

### Example YAML snippet

```yaml
wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password
```

### Anything in the logs that might be useful for us?

```txt
[08:12:32][D][wifi:497]: Found networks:
[08:12:32][I][wifi:541]: - '0619562e-bcbf-4bfc-97a8-b19e3c8f' [redacted]▂▄▆█
[08:12:32][D][wifi:542]:     Channel: 11
[08:12:32][D][wifi:543]:     RSSI: -76 dB
[08:12:32][I][wifi:541]: - '0619562e-bcbf-4bfc-97a8-b19e3c8f' [redacted]▂▄▆█
[08:12:32][D][wifi:542]:     Channel: 11
[08:12:32][D][wifi:543]:     RSSI: -69 dB
[08:12:32][I][wifi:541]: - '0619562e-bcbf-4bfc-97a8-b19e3c8f' [redacted]▂▄▆█
[08:12:32][D][wifi:542]:     Channel: 11
[08:12:32][D][wifi:543]:     RSSI: -58 dB
[08:12:38][I][wifi:313]: WiFi Connecting to '0619562e-bcbf-4bfc-97a8-b19e3c8f'...
[08:12:38][W][wifi:663]: WiFi network can not be found anymore.
[08:12:43][I][wifi:313]: WiFi Connecting to '0619562e-bcbf-4bfc-97a8-b19e3c8f'...
[08:12:43][W][wifi:663]: WiFi network can not be found anymore.
[08:12:43][I][wifi:313]: WiFi Connecting to '0619562e-bcbf-4bfc-97a8-b19e3c8f'...
[08:12:43][W][wifi:663]: WiFi network can not be found anymore.
[08:13:58][W][wifi_esp32:534][arduino_events]: Event: Disconnected ssid='0619562e-bcbf-4bfc-97a8-b19e3c8' reason='Probe Request Unsuccessful'
```

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by ssieb**
# What does this implement/fix?

https://github.com/esphome/esphome/pull/7668 broke 32 byte SSIDs.  This fixes them.

## Types of changes

- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- fixes https://github.com/esphome/issues/issues/6452

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [X] ESP32
- [X] ESP32 IDF
- [X] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [X] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/wifi/wifi_component_esp8266.cpp [hunk 1]
```diff
@@ -775,7 +783,11 @@ bool WiFiComponent::wifi_start_ap_(const WiFiAP &ap) {
     return false;
 
   struct softap_config conf {};
-  snprintf(reinterpret_cast<char *>(conf.ssid), sizeof(conf.ssid), "%s", ap.get_ssid().c_str());
+  if (ap.get_ssid().size() > sizeof(conf.ssid)) {
```
Comment (pzich) [inline esphome/components/wifi/wifi_component_esp8266.cpp:786]: I'm definitely rusty on pure C. Is `struct softap_config conf {};` guaranteed to be all nulls? Or does this need a `memset(&conf, 0, sizeof(conf));` as well?

Comment (ssieb) [reply esphome/components/wifi/wifi_component_esp8266.cpp:786]: For whatever reason, the ap struct has an ssid size field.  See the following lines.

**Comment by Pesc0** [comment]
I can confirm this fixes the issue

**Review by jesserockz — APPROVED** [review]

**Comment by jnawk** [comment]
Unit tests?

**Comment by ssieb** [comment]
esphome doesn't really have unit tests and I'm not sure how you would test this anyway.

**Comment by jnawk** [comment]
> esphome doesn't really have unit tests

That's unfortunate 

> I'm not sure how you would test this anyway.

Perhaps feed a 32 character SSID & see if the same 32 characters come back?   Fault behaviour is that esphome ends up looking for a SSID of only the first 31 characters - and if _that_ happens to exist in the environment (same security details etc), then everything works as if only 31 characters was given to start with.

**Pull request merged**